doc/source/development/rfc/rfc91_dataset_close.rst

```diff
@@ -0,0 +1,184 @@
+     * should be assumed as no longer being valid.
+     *
+     * If a driver implements this method, it should also call it from its
+     * dataset destructor.
+     *
+     * @since GDAL 3.7
+     */
+
+A typical implementation might look as the following:
+
+.. code-block:: c++
+
+    MyDataset::~MyDataset()
+    {
+       MyDataset::Close();
+    }
+
+    CPLErr MyDataset::Close()
+    {
+        CPLErr eErr = CE_None;
+        if( nOpenFlags != OPEN_FLAGS_CLOSED )
+        {
+            if( MyDataset::FlushCache(true) != CE_None )
+                eErr = CE_Failure;
+
+            // Do something driver specific
+            if (m_fpImage != nullptr)
+            {
+                if( VSIFCloseL(m_fpImage) != 0 )
+                {
+                    CPLError(CE_Failure, CPLE_FileIO, "VSIFCloseL() failed");
+                    eErr = CE_Failure;
+                }
+            }
+
+            // Call parent Close() implementation.
+            if( MyParentDatasetClass::Close() != CE_None )
+                eErr = CE_Failure;
+        }
+        return eErr;
+    }
+
+The default GDALDataset::Close() implementation sets nOpenFlags to OPEN_FLAGS_CLOSED
+
+
+C API
+-----
+
+GDALClose() and GDALFlushCache() are modified to return a CPLErr return value.
+
+Backward compatibility
+----------------------
+
+This is an ABI change that should not require more than rebuilding applications
+against GDAL headers.
+
+For out-of-tree drivers that implement GDALDataset::FlushCache(), they need to
+take into account the change in its signature.
+
+Out-of-tree drivers that have write capabilities are also encouraged to
+implement GDALDataset::Close().
+
+Limitations
+-----------
+
+Not all drivers will be modified to implement Close() in the candidate
+implementation, and even those modified might call internal methods of the
+driver that don't do error propagation. Consequently further work might be
+needed on a case-by-case to improve driver implementations.
+
+SWIG Bindings
+-------------
+
+The destructor of gdal.Dataset is modified to test the return value of GDALClose()
+and emits a CPLError(CE_Failure, ...) if the error state is clean (normally
+it should not)
+
+C/C++ command line utilities
+----------------------------
+
+C/C++ command line utilities are modified to test the return value of GDALClose()
+on output datasets, and return a non-zero return code for the process if GDALClose()
+returns an error.
+
+Testing
+-------
+
+autotest/pymod/gdaltest.py::testCreate() is modified to call FlushCache() and
+test its return value.
+
+The existing tests of the C/C++ command line utilities test GDALClose() modified
+behavior.
+
+Issues / pull requests
+----------------------
+
+Addresses https://github.com/OSGeo/gdal/issues/6886
+
+https://github.com/OSGeo/gdal/compare/master...rouault:gdal:dataset_FlushCache_return_CPLErr?expand=1 contains
+a candidate implementation.
+
+It updates all drivers that implements GDALDataset::FlushCache().
```

Comment:
Convert this paragraph to a list like this:

The candidate implementation does the following:

* implements GDALDataset::Close() for all drivers that derive from RawDataset.
* ...

Reply:
done